### PR TITLE
Python does not have distlib path in CentOS/RHEL.

### LINKED ISSUE #973: Python example fails with Gramine-Direct with recent hardening changes on CentOS/RHEL.
### Description of the problem

**Python example fails with Gramine-Direct with recent hardening changes on CentOS/RHEL.**

To make python work for Gramine in CentOS/RHEL, we initially removed distlib and apport from manifest files and changed the enclave size to 4GB
https://github.com/gramineproject/graphene/issues/2599

**_But with recent hardening changes, it fails, because of fs.mounts_**
Previously complete /usr folder was mounted, which has been now restrcited to /usr/arch_libdir

> `{ path = "/usr", uri = "file:/usr" }, replaced with { path = "/usr/{{ arch_libdir }}", uri = "file:/usr/{{ arch_libdir }}" },`

**With latest changes it is failing on RHEL with error "No module named six" and Numpy throws error for CentOS**

_We don't have docker setup right now for RHEL. In RHEL Baremetal Setup the "six" package is in /usr/local/lib which is not mounted now_
```
RHEL:
Numpy: /usr/lib64/python3.6/site-packages
Scipy: /usr/lib64/python3.6/site-packages
Six: /usr/local/lib/python3.6/site-packages
```

_Whereas for CentOS, numpy is getting installed in /usr/local/lib64 and scipy is /usr/local/lib which is also not mounted._
```
CentOS:
Numpy: /usr/local/lib64/python3.6/site-packages
Scipy: /usr/lib64/python3.6/site-packages
SIX: /usr/local/lib/python3.6/site-packages
```

So, in order to run Python with Gramine we need to make several changes. 
1. Remove distlib
2. Increase enclave size
3. Map the respective folder in fs.mount and sgx.trusted_files section

**Since we are officially supporting rpm packages now, we should also make sure all the examples in CI-Examples are compatible for Ubuntu, CentOS & RHEL**
https://gramine.readthedocs.io/en/stable/quickstart.html#install-gramine

> 	On RHEL-8-like distribution (like AlmaLinux 8, CentOS 8, Rocky Linux 8, …):
> 
> 	sudo curl -fsSLo /etc/yum.repos.d/gramine.repo https://packages.gramineproject.io/rpm/gramine.repo
> 	sudo dnf install gramine          # only the default, distro-provided kernel is supported

### Steps to reproduce

1. Checkout gramine package for CentOS/RHEL
2. make python example with **make SGX=1 ARCH_LIBDIR=/lib64**
3. Run make SGX=1 check

### Expected results

Python should work and no error should be seen

### Actual results

 **"No module named six" for RHEL and Numpy throws error for CentOS**

### Gramine commit hash

fd329ff42a0c5ce4edd171b2835c4e5796e90613

### PR DESCRIPTION
**Pull request by anjalirai-intel**
## Description of the changes 

To support Python for CentOS, RHEL & Ubuntu from Gramine, we need to replace distlib with numpy, scipy & six libraries path, because distlib path does not exist in CentOS & RHEL

Fixes #973. See it for context.

## How to test this PR? 
1. Make Python example using make SGX=1 on CentOS/RHEL
2. make SGX=1 check

- - -
This change is [Reviewable](https://reviewable.io/reviews/gramineproject/gramine/978)

[rhel_python3_sysconfig.txt](https://github.com/gramineproject/gramine/files/9798424/rhel_python3_sysconfig.txt)
[centos_python3_sysconfig.txt](https://github.com/gramineproject/gramine/files/9798426/centos_python3_sysconfig.txt)

### CODE REVIEW

**Review by dimakuv — CHANGES_REQUESTED** [review]
> Reviewed 2 of 2 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/gramineproject/gramine/978)* status: all files reviewed, 6 unresolved discussions, not enough approvals from maintainers (3 more required), not enough approvals from different teams (2 more required, approved so far: ), period found at the end of a one-liner (waiting on @anjalirai-intel)

*[a discussion](https://reviewable.io/reviews/gramineproject/gramine/978#-NEKVxpg3ym9AmQtTAAY:-NEKVxpg3ym9AmQtTAAZ:b-ncn45n) (no related file):*
This PR depends on https://github.com/gramineproject/gramine/pull/977. Blocking until that one is merged.

___
___
*[`-- commits` line 4 at r1](https://reviewable.io/reviews/gramineproject/gramine/978#-NEKW0JB0jBq73kCkwwH):*
@anjalirai-intel Please `git log` on the Gramine repo and check how we write commit messages. There are too many mistakes in this commit message to explain them one by one...
___
*[`CI-Examples/python/Makefile` line 5 at r1](https://reviewable.io/reviews/gramineproject/gramine/978#-NEKX1fa3yFpUYvR3y3C:-NEKX1fa3yFpUYvR3y3D:b-pl0u59) ([raw file](https://github.com/gramineproject/gramine/blob/db8167efc53ce6393c843a7b9b919a82d63703f5/CI-Examples/python/Makefile#L5)):*
> ```Makefile
> NUMPY_PATH ?= $(shell pip3 show numpy | grep Location: | cut -d" " -f2)/
> SCIPY_PATH ?= $(shell pip3 show scipy | grep Location: | cut -d" " -f2)/
> SIX_PATH ?= $(shell pip3 show six | grep Location: | cut -d" " -f2)/
> ```

I looked at the other scripts that we have for this Python example. Looks like indeed these three paths are enough to cover everything we do in the scripts. Looks good to me.
___
*[`CI-Examples/python/Makefile` line 5 at r1](https://reviewable.io/reviews/gramineproject/gramine/978#-NEKXGI6EZhBCvIEyrf-:-NEKXGI6EZhBCvIEyrf0:bk5zjc1) ([raw file](https://github.com/gramineproject/gramine/blob/db8167efc53ce6393c843a7b9b919a82d63703f5/CI-Examples/python/Makefile#L5)):*
> ```Makefile
> NUMPY_PATH ?= $(shell pip3 show numpy | grep Location: | cut -d" " -f2)/
> SCIPY_PATH ?= $(shell pip3 show scipy | grep Location: | cut -d" " -f2)/
> SIX_PATH ?= $(shell pip3 show six | grep Location: | cut -d" " -f2)/
> ```

Why do you add `/` at the end? I think these are not needed, please remove them.
___
*[`CI-Examples/python/python.manifest.template` line 11 at r1](https://reviewable.io/reviews/gramineproject/gramine/978#-NEKWZK-8fYk7KW-KEG4:-NEKWZK-8fYk7KW-KEG5:b-dhasnq) ([raw file](https://github.com/gramineproject/gramine/blob/db8167efc53ce6393c843a7b9b919a82d63703f5/CI-Examples/python/python.manifest.template#L11)):*
> ```
> 
> loader.insecure__use_cmdline_argv = true
> loader.pal_internal_mem_size="128M"
> ```

This won't be needed after #977 is merged.
___
*[`CI-Examples/python/python.manifest.template` line 19 at r1](https://reviewable.io/reviews/gramineproject/gramine/978#-NEKWK9y7FrRlQf27JNf:-NEKWK9y7FrRlQf27JNg:bx0newd) ([raw file](https://github.com/gramineproject/gramine/blob/3e461d607eb90de4c64190cca6910dbb40e7bf91/CI-Examples/python/python.manifest.template#L19)):*
> ```
>   { path = "/usr/{{ arch_libdir }}", uri = "file:/usr/{{ arch_libdir }}" },
>   { path = "{{ python.stdlib }}", uri = "file:{{ python.stdlib }}" },
>   { path = "{{ python.distlib }}", uri = "file:{{ python.distlib }}" },
> ```

Actually, you privately told me that `python.distlib` doesn't exist on CentOS/RHEL... So what is this path expanded to on these OSes? What is the error message that you get?

I'm surprised because `python.distlib` is supposed to be taken from OS settings. Or is it just an empty string?
___
*[`CI-Examples/python/python.manifest.template` line 34 at r1](https://reviewable.io/reviews/gramineproject/gramine/978#-NEKWcur9ARl-eZDcP74:-NEKWcur9ARl-eZDcP75:b-cam6vp) ([raw file](https://github.com/gramineproject/gramine/blob/db8167efc53ce6393c843a7b9b919a82d63703f5/CI-Examples/python/python.manifest.template#L34)):*
> ```
> sgx.debug = true
> sgx.nonpie_binary = true
> sgx.enclave_size = "4G"
> ```

Why? Please incorporate the changes from a related PR https://github.com/gramineproject/gramine/pull/938 into yours, and run the experiments again.

We definitely don't need 4GB of enclave memory. 1GB at maximum.

**Comment by anjalirai-intel** [comment]
@dimakuv i will fix the changes. But to answer your question about distlib

Distlib extends to /usr/lib64/python3/dist-packages. Firstly /usr/lib64/python3 path itself does not exist

Directory structure
 _**ls -al /usr/lib64/ | grep python**_
**lrwxrwxrwx.   1 root root        20 Oct 11  2019 libpython3.6m.so -> libpython3.6m.so.1.0
-rwxr-xr-x.   1 root root   3371672 Oct 11  2019 libpython3.6m.so.1.0
-rwxr-xr-x.   1 root root      7104 Oct 11  2019 libpython3.so
drwxr-xr-x.  32 root root      8192 Aug 30  2021 python3.6**



**_ls -al /usr/lib64/python3.6/ | grep dist_
drwxr-xr-x.   4 root root   4096 Aug 30  2021 distutils**

**Review by dimakuv — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/gramineproject/gramine/978)* status: all files reviewed, 6 unresolved discussions, not enough approvals from maintainers (3 more required), not enough approvals from different teams (2 more required, approved so far: ), period found at the end of a one-liner (waiting on @anjalirai-intel)

___
*[`CI-Examples/python/python.manifest.template` line 19 at r1](https://reviewable.io/reviews/gramineproject/gramine/978#-NEKWK9y7FrRlQf27JNf:-NEZrRzJEesPjPGbcZhC:b4qjrp6) ([raw file](https://github.com/gramineproject/gramine/blob/3e461d607eb90de4c64190cca6910dbb40e7bf91/CI-Examples/python/python.manifest.template#L19)):*
@anjalirai-intel Please reply via Reviewable next time. I'm copying your reply here, for the context (so that it's not like I'm talking to myself :))

> to answer your question about distlib
> Distlib extends to /usr/lib64/python3/dist-packages. Firstly /usr/lib64/python3 path itself does not exist [...]

Ok, I see the problem now: https://github.com/gramineproject/gramine/blob/9a4fd153b489dbfcad1112a6aae47c35f3503c5a/python/graminelibos/gen_jinja_env.py#L39-L42

Basically, we hard-code `distlib` to follow the conventions from Debian-based distros. We should fix this.

So where do we find `dist-packages/` on CentOS/RHEL (= RPM-based distros)? Or such a directory doesn't exist at all? Can you show the output of `python -m site` on CentOS and on RHEL?

**Comment by anjalirai-intel** [comment]
*[`CI-Examples/python/python.manifest.template` line 19 at r1](https://reviewable.io/reviews/gramineproject/gramine/978#-NEKWK9y7FrRlQf27JNf:-NEZtaEN3pN39G8C6LZ_:b4ksyk1) ([raw file](https://github.com/gramineproject/gramine/blob/3e461d607eb90de4c64190cca6910dbb40e7bf91/CI-Examples/python/python.manifest.template#L19)):*
<details><summary><i>Previously, dimakuv (Dmitrii Kuvaiskii) wrote…</i></summary><blockquote>

@anjalirai-intel Please reply via Reviewable next time. I'm copying your reply here, for the context (so that it's not like I'm talking to myself :))

> to answer your question about distlib
> Distlib extends to /usr/lib64/python3/dist-packages. Firstly /usr/lib64/python3 path itself does not exist [...]

Ok, I see the problem now: https://github.com/gramineproject/gramine/blob/9a4fd153b489dbfcad1112a6aae47c35f3503c5a/python/graminelibos/gen_jinja_env.py#L39-L42

Basically, we hard-code `distlib` to follow the conventions from Debian-based distros. We should fix this.

So where do we find `dist-packages/` on CentOS/RHEL (= RPM-based distros)? Or such a directory doesn't exist at all? Can you show the output of `python -m site` on CentOS and on RHEL?
</blockquote></details>

Sure. Basically we don't have dist-packages at all. This is the output from RHEL

sys.path = [
    '/home/intel/anjali/python_centos/CI-Examples/python',
    '/home/intel/gramine_install/usr/lib64/python3.6/site-packages',
    '/usr/lib64/python36.zip',
    '/usr/lib64/python3.6',
    '/usr/lib64/python3.6/lib-dynload',
    '/home/intel/.local/lib/python3.6/site-packages',
    '/usr/local/lib64/python3.6/site-packages',
    '/usr/local/lib/python3.6/site-packages',
    '/usr/lib64/python3.6/site-packages',
    '/usr/lib/python3.6/site-packages',
]
USER_BASE: '/home/intel/.local' (exists)
USER_SITE: '/home/intel/.local/lib/python3.6/site-packages' (exists)
ENABLE_USER_SITE: True

**Comment by anjalirai-intel** [comment]
*[`CI-Examples/python/python.manifest.template` line 19 at r1](https://reviewable.io/reviews/gramineproject/gramine/978#-NEKWK9y7FrRlQf27JNf:-NE_243s8Rf65qY6db0k:b-o01pdp) ([raw file](https://github.com/gramineproject/gramine/blob/3e461d607eb90de4c64190cca6910dbb40e7bf91/CI-Examples/python/python.manifest.template#L19)):*
<details><summary><i>Previously, dimakuv (Dmitrii Kuvaiskii) wrote…</i></summary><blockquote>

Thanks. Can you also show what `python -m sysconfig` outputs on CentOS and on RHEL? (Please show both outputs.)
</blockquote></details>

The output is very huge. I have attached file for both. Any specific thing do you want to look at it. I cannot attach huge files here in reviewable

**Review by dimakuv — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/gramineproject/gramine/978)* status: all files reviewed, 6 unresolved discussions, not enough approvals from maintainers (3 more required), not enough approvals from different teams (2 more required, approved so far: ), period found at the end of a one-liner (waiting on @anjalirai-intel)

___
*[`CI-Examples/python/python.manifest.template` line 19 at r1](https://reviewable.io/reviews/gramineproject/gramine/978#-NEKWK9y7FrRlQf27JNf:-NE_39Ma-K5aqUoukup1:b-6dkapl) ([raw file](https://github.com/gramineproject/gramine/blob/3e461d607eb90de4c64190cca6910dbb40e7bf91/CI-Examples/python/python.manifest.template#L19)):*
<details><summary><i>Previously, anjalirai-intel wrote…</i></summary><blockquote>

The output is very huge. I have attached file for both. Any specific thing do you want to look at it. I cannot attach huge files here in reviewable
</blockquote></details>

I'm interested in the `Paths:` output only. Sorry, should have mentioned. (And it should be not `python` but `python3` of course.)

**Comment by anjalirai-intel** [comment]
*[`CI-Examples/python/python.manifest.template` line 19 at r1](https://reviewable.io/reviews/gramineproject/gramine/978#-NEKWK9y7FrRlQf27JNf:-NE_3C9i8fTnn1adfY8Z:boq3btm) ([raw file](https://github.com/gramineproject/gramine/blob/3e461d607eb90de4c64190cca6910dbb40e7bf91/CI-Examples/python/python.manifest.template#L19)):*
<details><summary><i>Previously, dimakuv (Dmitrii Kuvaiskii) wrote…</i></summary><blockquote>

I'm interested in the `Paths:` output only. Sorry, should have mentioned. (And it should be not `python` but `python3` of course.)
</blockquote></details>

RHEL

python3 -m sysconfig
Platform: "linux-x86_64"
Python version: "3.6"
Current installation scheme: "posix_prefix"

Paths:
        data = "/usr"
        include = "/usr/include/python3.6m"
        platinclude = "/usr/include/python3.6m"
        platlib = "/usr/lib64/python3.6/site-packages"
        platstdlib = "/usr/lib64/python3.6"
        purelib = "/usr/lib/python3.6/site-packages"
        scripts = "/usr/bin"
        stdlib = "/usr/lib64/python3.6"

CentOS

[intel@ocsbesrhlrepo01 ~]$ python3 -m sysconfig
Platform: "linux-x86_64"
Python version: "3.6"
Current installation scheme: "posix_prefix"

Paths:
        data = "/usr"
        include = "/usr/include/python3.6m"
        platinclude = "/usr/include/python3.6m"
        platlib = "/usr/lib64/python3.6/site-packages"
        platstdlib = "/usr/lib64/python3.6"
        purelib = "/usr/lib/python3.6/site-packages"
        scripts = "/usr/bin"
        stdlib = "/usr/lib64/python3.6"

**Review by dimakuv — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/gramineproject/gramine/978)* status: all files reviewed, 6 unresolved discussions, not enough approvals from maintainers (3 more required), not enough approvals from different teams (2 more required, approved so far: ), period found at the end of a one-liner (waiting on @anjalirai-intel)

___
*[`CI-Examples/python/python.manifest.template` line 19 at r1](https://reviewable.io/reviews/gramineproject/gramine/978#-NEKWK9y7FrRlQf27JNf:-NE_4o0vBFz_BQpoBkf_:b-c27lpa) ([raw file](https://github.com/gramineproject/gramine/blob/3e461d607eb90de4c64190cca6910dbb40e7bf91/CI-Examples/python/python.manifest.template#L19)):*
<details><summary><i>Previously, anjalirai-intel wrote…</i></summary><blockquote>

RHEL

python3 -m sysconfig
Platform: "linux-x86_64"
Python version: "3.6"
Current installation scheme: "posix_prefix"

Paths:
        data = "/usr"
        include = "/usr/include/python3.6m"
        platinclude = "/usr/include/python3.6m"
        platlib = "/usr/lib64/python3.6/site-packages"
        platstdlib = "/usr/lib64/python3.6"
        purelib = "/usr/lib/python3.6/site-packages"
        scripts = "/usr/bin"
        stdlib = "/usr/lib64/python3.6"

CentOS

[intel@ocsbesrhlrepo01 ~]$ python3 -m sysconfig
Platform: "linux-x86_64"
Python version: "3.6"
Current installation scheme: "posix_prefix"

Paths:
        data = "/usr"
        include = "/usr/include/python3.6m"
        platinclude = "/usr/include/python3.6m"
        platlib = "/usr/lib64/python3.6/site-packages"
        platstdlib = "/usr/lib64/python3.6"
        purelib = "/usr/lib/python3.6/site-packages"
        scripts = "/usr/bin"
        stdlib = "/usr/lib64/python3.6"
</blockquote></details>

Ok, so to summarize the findings as I understand them:
- `distlib` is a concept that does **not** exist on RPM-based distros.
- Looking at the lists of Python paths above and the list of actual installation paths in #973, I see no correspondence.
- My conclusion: we should proceed with the new approach in this PR -- using `pip3 show <packagename>`.
- And for this PR to work nicely, we need to remove duplicates in sgx.trusted_files to decrease the final manifest size, see #977.

**Comment by anjalirai-intel** [comment]
*[`CI-Examples/python/python.manifest.template` line 34 at r1](https://reviewable.io/reviews/gramineproject/gramine/978#-NEKWcur9ARl-eZDcP74:-NE_8OyG2-ycyV_jek8H:b-5h6f81) ([raw file](https://github.com/gramineproject/gramine/blob/db8167efc53ce6393c843a7b9b919a82d63703f5/CI-Examples/python/python.manifest.template#L34)):*
<details><summary><i>Previously, dimakuv (Dmitrii Kuvaiskii) wrote…</i></summary><blockquote>

Why? Please incorporate the changes from a related PR https://github.com/gramineproject/gramine/pull/938 into yours, and run the experiments again.

We definitely don't need 4GB of enclave memory. 1GB at maximum.
</blockquote></details>

I incorporated those changes. It worked. I will upload new version

**Comment by anjalirai-intel** [comment]
*[`-- commits` line 4 at r1](https://reviewable.io/reviews/gramineproject/gramine/978#-NEKW0JB0jBq73kCkwwH):*
<details><summary><i>Previously, dimakuv (Dmitrii Kuvaiskii) wrote…</i></summary><blockquote>

@anjalirai-intel Please `git log` on the Gramine repo and check how we write commit messages. There are too many mistakes in this commit message to explain them one by one...
</blockquote></details>

Done.

**Comment by anjalirai-intel** [comment]
*[`CI-Examples/python/Makefile` line 5 at r1](https://reviewable.io/reviews/gramineproject/gramine/978#-NEKXGI6EZhBCvIEyrf-:-NE_fOVNDPBg8tQ8bHvU:b-896fix) ([raw file](https://github.com/gramineproject/gramine/blob/db8167efc53ce6393c843a7b9b919a82d63703f5/CI-Examples/python/Makefile#L5)):*
<details><summary><i>Previously, dimakuv (Dmitrii Kuvaiskii) wrote…</i></summary><blockquote>

Why do you add `/` at the end? I think these are not needed, please remove them.
</blockquote></details>

Done.

**Comment by anjalirai-intel** [comment]
*[`CI-Examples/python/python.manifest.template` line 11 at r1](https://reviewable.io/reviews/gramineproject/gramine/978#-NEKWZK-8fYk7KW-KEG4:-NE_fQGO6ugaf7AU1kNU:ba3dfob) ([raw file](https://github.com/gramineproject/gramine/blob/db8167efc53ce6393c843a7b9b919a82d63703f5/CI-Examples/python/python.manifest.template#L11)):*
<details><summary><i>Previously, dimakuv (Dmitrii Kuvaiskii) wrote…</i></summary><blockquote>

This won't be needed after #977 is merged.
</blockquote></details>

Removed

**Review by dimakuv — CHANGES_REQUESTED** [review]
> Reviewed 2 of 2 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/gramineproject/gramine/978)* status: all files reviewed, 3 unresolved discussions, not enough approvals from maintainers (3 more required), not enough approvals from different teams (2 more required, approved so far: ) (waiting on @anjalirai-intel)

___
*[`-- commits` line 4 at r1](https://reviewable.io/reviews/gramineproject/gramine/978#-NEKW0JB0jBq73kCkwwH):*
<details><summary><i>Previously, anjalirai-intel wrote…</i></summary><blockquote>

Done.
</blockquote></details>

Thanks. I will fix it slightly during final rebase:
- Add `Add CentOS/RHEL support ...` in the title.
- Add dots at the end of sentences.
- `path are` -> `paths are`

**Review by dimakuv — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/gramineproject/gramine/978)* status: all files reviewed, 4 unresolved discussions, not enough approvals from maintainers (3 more required), not enough approvals from different teams (2 more required, approved so far: ) (waiting on @anjalirai-intel)

*[a discussion](https://reviewable.io/reviews/gramineproject/gramine/978#-NEaegdJEAE6Ot5UYZDr:-NEbhtbz1oPyhUazGU5x:b-smzv9h) (no related file):*
I like this approach, but looking at my Ubuntu 20.04 system's paths:
```
>>> import sys; print('\n'.join(path for path in sys.path if path), end='')

/usr/lib/python38.zip
/usr/lib/python3.8
/usr/lib/python3.8/lib-dynload
/home/dimakuv/.local/lib/python3.8/site-packages
/usr/local/lib/python3.8/dist-packages
/usr/lib/python3/dist-packages
```

I see two things:
- Some dirs do not exist at all (`/usr/lib/python38.zip` doesn't exist), so they should be skipped alltogether in this Python helper.
- Some dirs are subpaths of other dirs, so they should be eliminated.

But in general this is a nice approach. @anjalirai-intel Can you verify that the list of paths that this helper Python script produces covers everything in CentOS and in RHEL?

___

**Comment by anjalirai-intel** [comment]
@dimakuv @woju 

This piece of code covers all python path in CentOS and RHEL

**RHEL**
> import sys; print('\n'.join(path for path in sys.path if path), end='')
/home/intel/gramine_install/usr/lib64/python3.6/site-packages
/usr/lib64/python36.zip
/usr/lib64/python3.6
/usr/lib64/python3.6/lib-dynload
/home/intel/.local/lib/python3.6/site-packages
/usr/local/lib64/python3.6/site-packages
/usr/local/lib/python3.6/site-packages
/usr/lib64/python3.6/site-packages
/usr/lib/python3.6/site-packages

**CentOS**
> import sys; print('\n'.join(path for path in sys.path if path), end='')
/home/intel/gramine_install/usr/lib64/python3.6/site-packages
/usr/lib64/python36.zip
/usr/lib64/python3.6
/usr/lib64/python3.6/lib-dynload
/home/intel/.local/lib/python3.6/site-packages
/usr/local/lib64/python3.6/site-packages
/usr/local/lib/python3.6/site-packages
/usr/lib64/python3.6/site-packages
/usr/lib/python3.6/site-packages

**Review by dimakuv — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/gramineproject/gramine/978)* status: all files reviewed, 4 unresolved discussions, not enough approvals from maintainers (3 more required), not enough approvals from different teams (2 more required, approved so far: ) (waiting on @anjalirai-intel and @woju)

*[a discussion](https://reviewable.io/reviews/gramineproject/gramine/978#-NEaegdJEAE6Ot5UYZDr:-NEdosxt8xLExO55vsWL:b-r2x2hp) (no related file):*
<details><summary><i>Previously, anjalirai-intel wrote…</i></summary><blockquote>

@dimakuv @woju 

This piece of code covers all python path in CentOS and RHEL

**RHEL**
> import sys; print('\n'.join(path for path in sys.path if path), end='')
/home/intel/gramine_install/usr/lib64/python3.6/site-packages
/usr/lib64/python36.zip
/usr/lib64/python3.6
/usr/lib64/python3.6/lib-dynload
/home/intel/.local/lib/python3.6/site-packages
/usr/local/lib64/python3.6/site-packages
/usr/local/lib/python3.6/site-packages
/usr/lib64/python3.6/site-packages
/usr/lib/python3.6/site-packages

**CentOS**
> import sys; print('\n'.join(path for path in sys.path if path), end='')
/home/intel/gramine_install/usr/lib64/python3.6/site-packages
/usr/lib64/python36.zip
/usr/lib64/python3.6
/usr/lib64/python3.6/lib-dynload
/home/intel/.local/lib/python3.6/site-packages
/usr/local/lib64/python3.6/site-packages
/usr/local/lib/python3.6/site-packages
/usr/lib64/python3.6/site-packages
/usr/lib/python3.6/site-packages
</blockquote></details>

@anjalirai-intel I think we can have an additional step in the Makefile of our Python example, where we will invoke the Python script (without Gramine, just natively) that lists these paths without duplicates and saves this list in a temporary file. And then the `python.manifest.file` includes the entries from this list.

Hm, this got complicated. If you're unsure how to implement this, me or Woju or someone else could try this out.

___

**Comment by anjalirai-intel** [comment]
<details><summary><i>Previously, dimakuv (Dmitrii Kuvaiskii) wrote…</i></summary><blockquote>

@anjalirai-intel I think we can have an additional step in the Makefile of our Python example, where we will invoke the Python script (without Gramine, just natively) that lists these paths without duplicates and saves this list in a temporary file. And then the `python.manifest.file` includes the entries from this list.

Hm, this got complicated. If you're unsure how to implement this, me or Woju or someone else could try this out.
</blockquote></details>

@woju @dimakuv You can look into this part. I am sure about manifest template changes

**Comment by anjalirai-intel** [comment]
<details><summary><i>Previously, anjalirai-intel wrote…</i></summary><blockquote>

@woju @dimakuv You can look into this part. I am sure about manifest template changes
</blockquote></details>

unsure*

**Review by dimakuv — CHANGES_REQUESTED** [review]
> Reviewed 2 of 2 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/gramineproject/gramine/978)* status: all files reviewed, 6 unresolved discussions, not enough approvals from maintainers (2 more required), not enough approvals from different teams (2 more required, approved so far: ) (waiting on @anjalirai-intel and @woju)

*[a discussion](https://reviewable.io/reviews/gramineproject/gramine/978#-NEpF27rD6IjsEGkVP8n:-NEpF27rD6IjsEGkVP8o:bekuzn) (no related file):*
You need to move changes in `gen_jinja_env.py` from #977  to here.

___
___
*[`CI-Examples/python/python.manifest.template` line 21 at r3](https://reviewable.io/reviews/gramineproject/gramine/978#-NEpEuTpEUciA_xJO7JD:-NEpEuTpEUciA_xJO7JE:bwmzu4g) ([raw file](https://github.com/gramineproject/gramine/blob/d9e530e4c15b2855ce8af7614ae671aac750dc2b/CI-Examples/python/python.manifest.template#L21)):*
> ```
>   { path = "/usr/{{ arch_libdir }}", uri = "file:/usr/{{ arch_libdir }}" },
>   { path = "{{ python.stdlib }}", uri = "file:{{ python.stdlib }}" },
>   {% for path in python.get_sys_path() %}
> ```

`get_sys_paths` (see my other comment in #977)
___
*[`CI-Examples/python/python.manifest.template` line 49 at r3](https://reviewable.io/reviews/gramineproject/gramine/978#-NEpF-OwAP1L_mmXMX_f:-NEpF-OwAP1L_mmXMX_g:bwmzu4g) ([raw file](https://github.com/gramineproject/gramine/blob/d9e530e4c15b2855ce8af7614ae671aac750dc2b/CI-Examples/python/python.manifest.template#L49)):*
> ```
>   "file:/usr/{{ arch_libdir }}/",
>   "file:{{ python.stdlib }}/",
>   {% for path in python.get_sys_path() %}
> ```

`get_sys_paths` (see my other comment in #977)

**Comment by anjalirai-intel** [comment]
*[`CI-Examples/python/python.manifest.template` line 21 at r3](https://reviewable.io/reviews/gramineproject/gramine/978#-NEpEuTpEUciA_xJO7JD:-NEpTLfmCkv83lmBbRjE:b-896fix) ([raw file](https://github.com/gramineproject/gramine/blob/d9e530e4c15b2855ce8af7614ae671aac750dc2b/CI-Examples/python/python.manifest.template#L21)):*
<details><summary><i>Previously, dimakuv (Dmitrii Kuvaiskii) wrote…</i></summary><blockquote>

`get_sys_paths` (see my other comment in #977)
</blockquote></details>

Done.

**Comment by anjalirai-intel** [comment]
<details><summary><i>Previously, dimakuv (Dmitrii Kuvaiskii) wrote…</i></summary><blockquote>

You need to move changes in `gen_jinja_env.py` from #977  to here.
</blockquote></details>

Done.

**Comment by anjalirai-intel** [comment]
*[`CI-Examples/python/python.manifest.template` line 49 at r3](https://reviewable.io/reviews/gramineproject/gramine/978#-NEpF-OwAP1L_mmXMX_f:-NEpTW1C8rJbHs01QzDD:bb6un1j) ([raw file](https://github.com/gramineproject/gramine/blob/d9e530e4c15b2855ce8af7614ae671aac750dc2b/CI-Examples/python/python.manifest.template#L49)):*
<details><summary><i>Previously, dimakuv (Dmitrii Kuvaiskii) wrote…</i></summary><blockquote>

`get_sys_paths` (see my other comment in #977)
</blockquote></details>

Done
